File: backend/src/presentation/controllers/message.controller.ts
import { Request, Response } from 'express';
import { inject, injectable } from 'inversify';
import { CreateMessageUseCase } from '../../application/use-cases/create-message.use-case';
import { TYPES } from '../../types';

@injectable()
export class MessageController {
  constructor(
    @inject(TYPES.CreateMessageUseCase) private createMessageUseCase: CreateMessageUseCase
  ) {}

  async create(req: Request, res: Response): Promise<Response> {
    try {
      const { text, authorId, channelId } = req.body;
      const message = await this.createMessageUseCase.execute({ text, authorId, channelId });
      return res.status(201).json(message);
    } catch (error) {
      return res.status(500).json({ error: 'Internal server error' });
    }
  }
}

File: backend/presentation/controllers/message_controller.go
package controllers

import (
	"backend/application/usecases"
	"backend/domain"
	"errors"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/go-playground/validator/v10"
)

type MessageController struct {
	CreateMessageUseCase *usecases.CreateMessageUseCase
	Validate             *validator.Validate
}

func NewMessageController(createMessageUseCase *usecases.CreateMessageUseCase, validate *validator.Validate) *MessageController {
	return &MessageController{
		CreateMessageUseCase: createMessageUseCase,
		Validate:             validate,
	}
}

func (ctrl *MessageController) CreateMessage(c *gin.Context) {
	var input usecases.CreateMessageInputDTO
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	if err := ctrl.Validate.Struct(input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	message, err := ctrl.CreateMessageUseCase.Execute(input)
	if err != nil {
		switch {
		case errors.Is(err, domain.ErrInvalidInput):
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		case errors.Is(err, domain.ErrConflict):
			c.JSON(http.StatusConflict, gin.H{"error": err.Error()})
		default:
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create message"})
		}
		return
	}

	c.JSON(http.StatusCreated, message)
}